damnit/backend/extraction_control.py

```diff
@@ -0,0 +1,247 @@
+"""Machinery to run Slurm jobs for extracting data
+
+See extract_data.py for what happens inside the jobs this launches.
+"""
+import getpass
+import logging
+import os
+import shlex
+import subprocess
+import sys
+from contextlib import contextmanager
+from ctypes import CDLL
+from dataclasses import dataclass
+from pathlib import Path
+from threading import Thread
+
+from extra_data.read_machinery import find_proposal
+
+from .db import DamnitDB
+from ..context import RunData
+
+log = logging.getLogger(__name__)
+
+
+# Python innetgr wrapper after https://github.com/wcooley/netgroup-python/
+def innetgr(netgroup: bytes, host=None, user=None, domain=None):
+    libc = CDLL("libc.so.6")
+    return bool(libc.innetgr(netgroup, host, user, domain))
+
+
+def default_slurm_partition():
+    username = getpass.getuser().encode()
+    if innetgr(b'exfel-wgs-users', user=username):
+        return 'exfel'
+    elif innetgr(b'upex-users', user=username):
+        return 'upex'
+    return 'all'
+
+
+def process_log_path(run, proposal, ctx_dir=Path('.'), create=True):
+    p = ctx_dir.absolute() / 'process_logs' / f"r{run}-p{proposal}.out"
+    if create:
+        p.parent.mkdir(exist_ok=True)
+        if p.parent.stat().st_uid == os.getuid():
+            p.parent.chmod(0o777)
+        p.touch(exist_ok=True)
+        if p.stat().st_uid == os.getuid():
+            p.chmod(0o666)
+    return p
+
+
+@contextmanager
+def tee(path: Path):
+    with path.open('ab') as fout:
+        r, w = os.pipe()
+        def loop():
+            while b := os.read(r, 4096):
+                fout.write(b)
+                sys.stdout.buffer.write(b)
+                sys.stdout.flush()
+
+        thread = Thread(target=loop)
+        thread.start()
+        try:
+            yield w
+        finally:
+            os.close(w)
+            thread.join()
+            os.close(r)
```

Comment:
For the file, it will be flushed when we close it on leaving the with block. It's [documented](https://docs.python.org/3/library/io.html#io.IOBase.close) that the close method flushes, and it's a convention that leaving a `with` block for an open file is equivalent to that.

The way this is used, we don't write to the pipe in this process, but pass it to a subprocess. So flushing the write end of the pipe in this process wouldn't have any effect, because it's not our buffer that matters. Any remaining data should be flushed as the child process exits.